### PR TITLE
Add Galaxy Notification System

### PR DESCRIPTION
**Pull request by davelopez**
Builds on (and supersedes) the work started in #14041

This is an overhaul of the originally designed `Notification System` to support more complex scenarios and features described in #1023 and #2324

# Features

### Notification Scheduling

Notifications can be scheduled to be displayed to the users on a future date by specifying a `publication_time`.

### Notification Expiration

Notifications will have a default expiration date and will be removed permanently from the database regularly after their expiration date is over. Either the `cleanup` scripts or a cron-like celery task can regularly remove expired notifications to avoid bloating the DB.

### Notification Broadcasting

A special type of notification meant to support server-wide announcements. These notifications will be usually treated differently than regular notifications and will be displayed in the main view as banners or alerts for every user (authenticated or anonymous).
Some use cases for these notifications can be:

- Server scheduled maintenance notifications.
- Server high load warnings.
- Event announcements.
- User agreements (cookies, privacy policy, etc.).

### Structured Notifications

Notifications can have different categories and _variants_ and their content (JSON) is well-defined depending on the category. In general, the notification may just contain the necessary metadata to build the actual notification message in the client using the appropriate localization and display style.

#### Possible notification categories:

- `message`: A direct message to the user.
- `new_shared_item`: new resource (history, workflow, visualization or page) shared with the user, can contain the link to the shared resource.
- `workflow_execution_completed`: There is no hook yet for this event, but it will be possible in the future.
- More ideas here...

#### User Preferences

The user can decide **what** and **how** notifications will be displayed by opting out of specific notification categories or channels.
All notifications will be displayed in a dedicated panel in the Galaxy UI unless completely opt-out, in addition, the user can decide on which channels the notification can be sent.

Possible notifications channels un/selectable by the user for each notification category:

- push: This will display a web push notification
- email: The notification can be optionally sent by email (not included in this PR)
- matrix: Direct message via Matrix.org (not included in this PR)

These notification channels are just an idea for now and their implementations may or may not be included in this PR.

#### Favorites

~~The user may mark some notifications as `favorite`. This will prevent the notification from automatically expiring and offers more filtering options. This might be interesting, for example, to save a notification with useful information or links for future use.~~
EDIT: This has been removed.

#### Deletion

The user can always manually delete the notification without having to wait for the expiration date.

# Backend Changes

## Database models

The database models have been extended to include more information and support the above features:

```mermaid
erDiagram
    NOTIFICATION ||--o{ USER_NOTIFICATION_ASSOCIATION : notifies_users
    NOTIFICATION {
        int id PK
        datetime create_time "The creation date."
        datetime update_time "The last time the notification content was updated."
        datetime publication_time "The date of publication, can be a future date to allow scheduling."
        datetime expiration_time "The expiration date, expired notifications will be permanently removed from DB regularly."
        string source "Who (or what) generated the notification."
        string category "Category of the notification, defines its contents. Used for filtering, un/subscribing, etc."
        string variant "Defines the 'importance' of the notification. Used for filtering, highlight rendering, etc."
        json content "Structured content in JSON. Depending on the category of the notification."

    }
    USER_NOTIFICATION_ASSOCIATION {
        int id PK
        int user_id FK
        int notification_id FK
        datetime seen_time "Date when the user 'opened' the notification (or null if unread)."
        bool deleted "Marking a notification as deleted will hide it from the user until it expires. Can be undeleted."
    }
```

## API

The rest API has been also extended to support the features and use cases described above.

Access the interactive API documentation here: http://127.0.0.1:8080/api/docs#/notifications

image

## TODO
- [x] Use the user preferences to skip notification creation
   - [ ] Create associations in an external task when targeting a potentially large number of users? Let's see how it behaves and then follow-up in a different PR.
- [x] Add expiration/deletion cleanup mechanism
- [x] Add notification hooks for shared items (histories, workflows, etc.)
- [x] Add config option to globally enable/disable notifications
- [x] Add test coverage

# Frontend Changes

There are two main kinds of notifications, *User notifications* and *server-wide (or broadcast) notifications*.

## User notifications

User notifications are specific for each user and can be marked as read or deleted by them. In this PR two sub-types of user notifications are implemented, **Shared Items** and general **Message** notifications.

image

Notification preferences allow you to subscribe/unsubscribe from certain types of User Notifications and choose through which channels you want to be notified (currently only the *Push* notification channel is available others will be added in the future).

image

## Broadcast notifications (server-wide)

There are again two sub-types of broadcasted notifications, depending on the *variant*. "regular" broadcast (info and warning variant) can be used to show Terms of use, Cookies policies, and all sorts of server-wide messages, while **urgent** variants are a bit more intrusive and require a bit more attention from the user.

### Regular broadcasts example
Screenshot from 2023-06-16 16-04-17

### Urgent variant example

image

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/schema/notifications.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    """These notification categories can be opt-out by the user and will be
+    displayed in the notification preferences.
+    """
+
+    message = "message"
+    near_quota_limit = "near_quota_limit"
+    new_history_shared = "new_history_shared"
+    new_workflow_shared = "new_workflow_shared"
+    new_page_shared = "new_page_shared"
+    new_visualization_shared = "new_page_shared"
```
Comment (mvdbeek) [inline lib/galaxy/schema/notifications.py]:
```suggestion
    new_visualization_shared = "new_visualization_shared"
```
?

File: lib/galaxy/schema/notifications.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    workflow_execution_completed = "workflow_execution_completed"
```
Comment (mvdbeek) [inline lib/galaxy/schema/notifications.py]:
```suggestion
    workflow_execution_completed = "workflow_execution_completed"
```
😍 I'd love to build the backend for that, just be aware we don't have a way to hook into that yet

File: lib/galaxy/schema/notifications.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+
+class NotificationContentBase(Model):
+    subject: str
+    message: str  # markdown?
+
+
+class ActionLink(Model):
+    action_name: str
+    link: AnyUrl
```
Comment (mvdbeek) [inline lib/galaxy/schema/notifications.py]: Can we not do urls, but return entity ids, so the client can build whatever operation is appropriate for an entity ?

So if it's a workflow_id, you could view, edit, run, download etc ... ?

Take a look at `WarningReason`, `FailureReason`, `CancelReason` ? That way we also don't need to hook into the backend routing for generating routes

File: lib/galaxy/schema/notifications.py (hunk 1 above)
Comment (davelopez) [reply lib/galaxy/schema/notifications.py]: Ups copy&paste is ruled by the devil :sweat_smile: Thanks!

File: lib/galaxy/schema/notifications.py (hunk 2 above)
Comment (davelopez) [reply lib/galaxy/schema/notifications.py]: Ahh good to know, those categories are just potential ideas for the moment, I'll disable that one for now.

File: lib/galaxy/schema/notifications.py (hunk 3 above)
Comment (davelopez) [reply lib/galaxy/schema/notifications.py]: Sure, that was the idea. Have a look at `WorkflowExecutionCompletedNotificationContent` below.

This `ActionLink` is only used in *Broadcasted notifications*, which are generic, and this is a naive attempt to support "Actions" or "Buttons" that can register specific actions.

For example, imagine creating a survey that gets displayed to every user (as a broadcasted notification), and there are a couple of options, those options could be links to a potential API endpoint (inside or outside Galaxy) that register the response. 
Or another example is when you click "Agree" on a privacy policy or cookie agreement. There would be an `agree` action link that points somewhere to read the terms, etc... But maybe it is too generic... :thinking: 

Those action links are supposed to be written by the admin or the person in charge of creating those special notifications, so they are not generated by Galaxy.

Comment (mvdbeek) [reply lib/galaxy/schema/notifications.py]: Ah, great, for outside links that makes a lot of sense.

Comment (davelopez) [reply lib/galaxy/schema/notifications.py]: Ohh wait, I see what you mean, the `share_link` below should go :sweat_smile:

File: lib/galaxy/model/migrations/alembic/versions_gxy/f6fb225c05ab_add_notification_table.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+"""Add Notification Table
```
Comment (jdavcs) [inline lib/galaxy/model/migrations/alembic/versions_gxy/f6fb225c05ab_add_notification_table.py]: Could you, please, move the revision script into its own commit? I haven't seen a case (yet) when we'd need the revision commit without the corresponding model change commit, but that was one of the "lessons learned" items in a 2014 PyCon Alembic talk, and I do think it's a good approach (if anything, that's more granular commits, which may be convenient).

File: test/unit/data/model/mapping/test_gxy_model_mapping.py [hunk 5]
```diff
@@ -1317,6 +1317,84 @@ def test_relationships(self, session, cls_, group, role):
             assert stored_obj.role.id == role.id
```
Comment (jdavcs) [inline test/unit/data/model/mapping/test_gxy_model_mapping.py]: We are dropping all mapping tests. It's not merged yet only because it got stuck in a large SA2.0 PR which is not yet ready for review. There's a more detailed explanation in the backend meeting minutes from 1/31/23 (see  bullet point 1.3). So this file will be dropped. (sorry for any wasted time - I should've dropped those already)

File: lib/galaxy/schema/notifications.py (hunk 2 above)
Comment (jmchilton) [reply lib/galaxy/schema/notifications.py]: near_quota_limit should also be disabled for now I assume (push back if I'm wrong). We present a quota bar to the user in the UI - the client could just look at this and decide what to do - rather than having to check and fire this every time we write any object in Galaxy. 

This is overall wonderful. I'd love to see the notifications implemented - the sharing stuff seems like a good start. There is also a job completion -> email hook in Galaxy that you can probably piggy back off of to build another useful thing pretty quickly.

File: lib/galaxy/model/__init__.py [hunk 6]
```diff
@@ -2457,6 +2458,46 @@ def __init__(self, user, group):
...
+    __tablename__ = "notification"
+
+    id = Column(Integer, primary_key=True)
+    create_time = Column(DateTime, default=now)
+    update_time = Column(DateTime, default=now, onupdate=now)
+    publication_time = Column(DateTime, default=now)
+    expiration_time = Column(DateTime, default=now() + timedelta(days=30 * 6))
+    source = Column(String(32), index=True)
+    category = Column(String(64), index=True)
+    variant = Column(String(16), index=True)
```
Comment (jdavcs) [inline lib/galaxy/model/__init__.py]: maybe add very brief inline comments (on the right?) explaining what there are? Like what you have in the data models table in the description of this PR. That would be very helpful for anyone reading the model definition code down the road.

Comment (jdavcs) [inline lib/galaxy/model/__init__.py]: Does the `variant` field imply a scale? ("Defines the 'importance' of the notification.") If so, maybe use an integer instead? Not only will that give us more flexibility (avoid the need for "high" vs. "somewhat high" vs. "very very high", etc.), we also get natural ordering.

File: lib/galaxy/model/__init__.py [hunk 7]
```diff
@@ -2457,6 +2458,46 @@ def __init__(self, user, group):
...
+    __tablename__ = "user_notification_association"
+
+    id = Column(Integer, primary_key=True)
+    user_id = Column(Integer, ForeignKey("galaxy_user.id"), index=True)
+    notification_id = Column(Integer, ForeignKey("notification.id"), index=True)
+    seen_time = Column(DateTime, nullable=True)
+    favorite = Column(Boolean, index=True, default=False)
+    deleted = Column(Boolean, index=True, default=False)
+
+    user = relationship("User", back_populates="all_notifications")
```
Comment (jdavcs) [inline lib/galaxy/model/__init__.py:2493]: You need to add the `all_notifications` relationship definition to the `User` class (same mirrored `back_populates` as well).

File: lib/galaxy/model/unittest_utils/gxy_model_fixtures.py [hunk 8]
```diff
@@ -895,6 +895,17 @@ def make_instance(*args, **kwds):
     return make_instance
```
Comment (jdavcs) [inline lib/galaxy/model/unittest_utils/gxy_model_fixtures.py]: Both fixtures can be dropped (together with the mapping tests) - again, sorry about that!

File: test/unit/data/model/mapping/test_gxy_model_mapping.py (hunk 5 above)
Comment (davelopez) [reply test/unit/data/model/mapping/test_gxy_model_mapping.py]: No worries! I had a vague feeling/memory this was the case but since the tests were already there in the original PR... I didn't think of it too much :man_shrugging: 

I'll drop them :+1: Thank you!

File: lib/galaxy/schema/notifications.py (hunk 2 above)
Comment (davelopez) [reply lib/galaxy/schema/notifications.py]: Thank you! That makes a huge lot of sense @jmchilton!

I'll look into the job completion hook :+1:

File: lib/galaxy/model/__init__.py (hunk 6 above)
Comment (davelopez) [reply lib/galaxy/model/__init__.py]: The idea was more along the line of something like "info", "warning", "urgent", etc. that can be used to render or highlight the notifications differently (colors, icons, etc.). I think the ordering should be always by date, but this `variant` can be used for filtering, i.e. display only "urgent" notifications.
Of course, these enums can be mapped as integers in the DB if this is more desirable, I just thought they are more readable as strings, but I'm happy to change that if this is more optimum :)

Comment (jdavcs) [reply lib/galaxy/model/__init__.py]: Ah, I see. Those values are more informative than a simple integer scale - let's keep them as strings then.
With regard to the database, it's all the same here. Thanks for the explanation!

**Comment by bgruening** [comment]
@itisAliRH thanks for working on this. It would be super nice if we can have the notification framework in the next release.

I'm a bit sad to see the "save notification" / "favorite" feature being removed. @hexylena and me had some use-cases for it and its github has a similar feature. Maybe we should have called it "save" and not "favorite" to better describe the feature.

**Comment by dannon** [comment]
@hexylena In our discussion, the main difference was that the types of notifications Galaxy will be sending seemed, to the group, fundamentally different from the Github notifications that were brought up as a canonical example of a platform that features useful 'favorite' notifications.  It seemed like we could definitely get away with an ACK/NACK and that would be sufficient, and would clean up and simplify the interface significantly.

That said, I should have made sure conversation made it to this PR after the meeting, good point -- we'll do better in the future to bridge the gap between folks who have the bandwidh to attend the WG and those who do not.  On the other side of the coin -- it sounds like you and @bgruening have had discussions and planning around this feature with specific use cases in mind -- can you point to somewhere those are outlined?  That'd be helpful, and I wish we'd known about it before.

**Comment by jdavcs** [comment]
@itisAliRH: most of the session-related test failures are happening because of code like this: `with self.sa_session.begin():`. This attempts to start a new transaction. With `autocommit=True` this worked because the only times a session was in a transaction was during a flush, when a transaction was started and committed implicitly. With `autocommit=False` (current setting), a transaction happens independently of the flush statement. A session is initially not in transaction mode, but as soon as there is a need for a transaction (an object whose state is managed by the session was modified, or a query was submitted to the database, explicitly or due to a lazy load, etc.), a transaction is created. Trying to start a new transaction on top of it will raise an error.

What to do: we shouldn't use `session.begin()`. Instead, we add a commit where we would normally have a flush (or at the end of the code block that would go inside the `session.begin` context manager). So, instead of this:
```py
with session.begin():
    session.execute(whatever)
```
we do this:
```py
session.execute(whatever)
with transaction(session):
    session.commit()
```
The `transaction` context manager is imported from `galaxy.model.base`. It ensures that if there is no transaction, it will be created.

I'll make the edits and will push to your branch directly, I hope you don't mind.

**Comment by jdavcs** [comment]
Database revisions (aka migrations) should be, ideally, placed in their own commit. If a revision script needs an edit, that edit needs to go into the same commit (so, we need to rebase). Otherwise (if the previous version has been merged), we need a new revision script. So, this commit: 4625293a0f6bb20f47fb1e91dfd1f2da97c1f119 must be edited: the change to the revision script should be extracted and combined with 3f678b841db1eea1d4347bd5de4f9f951f5b5eb4. So this requires a rebase.

**Comment by jdavcs** [comment]
The `test/unit/app/managers/test_NotificationManager.py::TestUserNotifications::test_cleanup_expired_notifications` failure is legitimate. The test broke here: 4625293a0f6bb20f47fb1e91dfd1f2da97c1f119. The key difference is [here](https://github.com/galaxyproject/galaxy/pull/15663/files#diff-06d9d006eb4b4e437413bc0da2b39b4293c2dd0cc5c12f70b59e912dad710408R328). The `rowcount` value is 0, but should be 1 (prior to this commit). That value is the number of matched rows in the WHERE clause (which is the `non_expired_notification_ids` query - if you trace it, you'll find that the one obvious change is the removal of the `favorite` attribute - which may or may not be the root cause).

**Comment by hexylena** [comment]
@dannon thank you, that would be very helpful.

Essentially all of my discussion and planning was on the original issue where I proposed this feature 8 years ago (https://github.com/galaxyproject/galaxy/issues/1023) and in more detail 7 years ago (https://github.com/galaxyproject/galaxy/issues/2324 since I'd forgotten about the first). But I can't seem to find them linked, neither from this PR, nor the last one but maybe I'm missing it in a collapsed comment?

I was under the impression these PRs were targeting the use cases @bgruening and I laid out in those two old issues, but maybe someone independently decided it was a good idea and never was aware of all of our public discussions and planning around the topic? That could explain some of my impressions, if this was ui-ux driven, rather than centered around our planning.

See especially the list https://github.com/galaxyproject/galaxy/issues/2324#issuecomment-910341495

Cases of notifications that might benefit from "later", taken from #2324, include

- New dataset was uploaded for you in this data library
- Approaching storage quota
- New features added in Galaxy that we would like to make users aware

**Comment by bgruening** [comment]
> I was under the impression these PRs were targeting the use cases @bgruening and I laid out in those two old issues

Yes it was. We wanted to create a flexible system that is able to address a lot of the collected use-cases.

> the main difference was that the types of notifications Galaxy will be sending seemed, to the group, fundamentally different from the Github notifications

Maybe that is the misunderstanding. The API allows sending notifications for all kind of events and at least EU is planning to use this feature a lot to communicate with users. The events that "Galaxy sents" in this PR is nowhere complete and I would expect they will grow over time. But even more importantly - its not Galaxy alone that will send events, it will be admins using bioblend (work is in progress for that) sending all kinds of notifications for different use cases.

**Comment by hexylena** [comment]
> Yes it was. We wanted to create a flexible system that is able to address a lot of the collected use-cases.

Ah if those were part of the consideration, then I guess we should add "Fixes #1023" and "Fixes #2324" to the parent, I am very excited to see them addressed.

**Comment by dannon** [comment]
Again taking the role of trying to summarize things, and trying to figure out the best path forward with the freeze in mind, for speed I've talked to I think everyone involved here out of band.  To get this in for the release we're going to leave this as the streamlined, slightly simpler interface discussed in the meeting and implemented here in subsequent commits, and follow up after.  

If we *do* decide to add 'pinned' or 'saved' notifications later, that's easy enough to add on with what we've already written -- the  timing complications here were just unlucky with individual holidays, the freeze, GCC, etc., but I think this will be fine and we can continue to iterate on things to keep improving it to where everyone's happy.

It'll be great to have proper notifications in Galaxy in 23.1!

**Comment by jdavcs** [comment]
This should fix the last test failures. (see commit description for details on changes). 
@itisAliRH marking this ready for review - I assume it's good to go?

**Comment by hexylena** [comment]
> If we do decide to add 'pinned' or 'saved' notifications later

Yes, please, two admins/power users have use cases and would like it back :) Would you like me to file an issue tracking this, to request it back (in the next version)?

**Comment by davelopez** [comment]
@jdavcs, Thank you so much for helping Alireza and fixing the db transactions while I was away, and sorry it was a bumpy road with the rebase.

@hexylena, My bad, I should have linked those issues in this PR, I definitely designed the Notification System backend taking all those into account, I just forgot to link them :sweat_smile:. I will update the PR description with the links and all the information I can, including some examples of the UI that Alireza has been working on.

In this first version, not every single sub-feature will be implemented, but rather the foundation will be there to make it easy to extend. In summary, 3 types of notifications are already implemented in this PR:
-  *Shared Item* notifications: for now, these are the only "automatic" notifications, so the target user will get a notification with a link to the resource (history, page, visualization, etc.) that was shared.
- Regular *Message" notifications: these can contain any markdown content and can be used to notify individual users, groups or roles, or a combination of them. They can only be created/sent by admins, the UI for this will be added in a follow-up PR so for the 23.1 version only the API can be used to create them.
- Broadcast notifications: these will be basically addressing #1023 but always server-wide for all users, it should be flexible enough for handling server maintenance notices to cookies or terms agreement notices, etc., and supports markdown too.

**Review by jdavcs — APPROVED** [review]
looks great!

**Review by dannon — APPROVED** [review]

**Comment by dannon** [comment]
@davelopez Thanks so much for taking on this project and pushing it forward!

**Comment by itisAliRH** [comment]
Big thanks to everyone involved for getting our pull request merged! Great teamwork! 🙌
